File: cofferdam_app/cofferdam_app/page/cofferdam_status/cofferdam_status.py
"""Controller for the Cofferdam Status page (read-only operational dashboard).

All data assembly happens here; the JS layer only renders. Every method is
whitelisted and gated to System Manager.

Design constraint (see docs/status-page-plan.md): this module must never import
``cofferdam`` (or ``cofferdam_app.policy``, which imports it) at module scope.
``cofferdam_app`` can be installed and running while the ``cofferdam`` library is
absent — the hooks import it lazily — so the page must render *despite* a missing
library in order to report that fact. All library access goes through
``_load_library()`` inside a try/except.
"""

from __future__ import annotations

import os
from datetime import UTC, datetime
from typing import Any

import frappe

_ROLE = "System Manager"

# Interception coverage table. The two live rows correspond to the doc_events
# wired in hooks.py; the rest are planned.
_COVERAGE: list[dict[str, Any]] = [
    {"path": "frappe.sendmail() → Email Queue", "intercepted": True, "note": "before_insert hook"},
    {
        "path": "Frappe Webhook → Webhook Request Log",
        "intercepted": True,
        "note": "before_insert hook (confirmed on v16)",
    },
    {"path": "ERPNext Slack Webhook", "intercepted": False, "note": "Planned"},
    {"path": "ERPNext payment gateways", "intercepted": False, "note": "Planned"},
    {"path": "ERPNext shipping carriers", "intercepted": False, "note": "Planned"},
    {"path": "ERPNext e-commerce connectors", "intercepted": False, "note": "Planned"},
]


def _load_library() -> tuple[dict[str, Any] | None, str | None]:
    """Lazily import cofferdam and the app policy helpers.

    Returns ``(bundle, None)`` on success or ``(None, error_message)`` if the
    ``cofferdam`` library is not importable. ``cofferdam_app.policy`` imports
    ``cofferdam`` at its top, so a missing library surfaces here as ImportError.
    """
    try:
        import cofferdam

        from cofferdam_app import policy as app_policy
    except ImportError as exc:
        return None, f"cofferdam library not importable: {exc}"
    return {"cofferdam": cofferdam, "app_policy": app_policy}, None


def _app_version() -> str:
    try:
        import cofferdam_app

        return getattr(cofferdam_app, "__version__", "unknown")
    except Exception:  # pragma: no cover - the app is always importable here
        return "unknown"


def _abs_policy_path() -> str:
    """Absolute path to this site's environment_policy.toml."""
    return os.path.abspath(frappe.get_site_path("environment_policy.toml"))


@frappe.whitelist()
def get_data() -> dict[str, Any]:
    """Assemble the full dashboard payload. Never raises for a missing library."""
    frappe.only_for(_ROLE)
    site: str = frappe.local.site
    errors: list[str] = []

    lib, lib_err = _load_library()
    library_installed = lib is not None
    if lib_err:
        errors.append(lib_err)

    data: dict[str, Any] = {
        "library_installed": library_installed,
        "site": site,
        "environment": None,
        "policy_file_path": None,
        "policy_file_exists": False,
        "policy_file_mtime": None,
        "policy_loaded_in_this_worker": False,
        "library_version": None,
        "app_version": _app_version(),
        "default_decision": None,
        "mail": None,
        "integrations": [],
        "effects": [],
        "credentials": [],
        "coverage": list(_COVERAGE),
        "errors": errors,
    }

    if lib is None:
        return data

    cofferdam = lib["cofferdam"]
    app_policy = lib["app_policy"]

    data["library_version"] = getattr(cofferdam, "__version__", None)
    data["policy_file_path"] = app_policy.policy_path(site)
    data["policy_loaded_in_this_worker"] = site in app_policy._policy_cache

    abs_path = _abs_policy_path()
    if os.path.exists(abs_path):
        data["policy_file_exists"] = True
        mtime = os.stat(abs_path).st_mtime
        data["policy_file_mtime"] = datetime.fromtimestamp(mtime, tz=UTC).isoformat()
    else:
        errors.append(f"policy file not found at {data['policy_file_path']}")
        return data

    # Load fresh from disk for display (this is not the hot path). The cache
    # status above is reported separately and truthfully.
    try:
        policy = cofferdam.load_policy(abs_path)
    except Exception as exc:
        errors.append(f"policy load failed: {exc}")
        return data

    _populate_from_policy(data, policy)
    return data


def _populate_from_policy(data: dict[str, Any], policy: Any) -> None:
    """Fill the display fields from a parsed Policy. No secrets are emitted."""
    data["environment"] = policy.environment.value
    data["default_decision"] = policy.default_decision.value

    if policy.mail is not None:
        data["mail"] = {
            "mode": policy.mail.mode,
            "sink": policy.mail.sink,
            "allow_domains": list(policy.mail.allow_domains),
            "decorate": policy.mail.decorate,
        }

    for name, integ in policy.integrations.items():
        data["integrations"].append(
            {
                "name": name,
                "kind": integ.kind.value,
                "enabled": integ.enabled,
                "credential": integ.credential,
                "allowed_hosts": list(integ.allowed_hosts),
                "allowed_methods": list(integ.allowed_methods),
                "allowed_operations": list(integ.allowed_operations),
                "allow_authorize": integ.allow_authorize,
                "allow_capture": integ.allow_capture,
            }
        )

    for kind, scopes in policy.effects.items():
        for scope, rule in scopes.items():
            data["effects"].append(
                {
                    "kind": kind,
                    "scope": scope,
                    "enabled": rule.enabled,
                    "allow_domains": list(rule.allow_domains),
                }
            )

    for name, cred in policy.credentials.items():
        secret_env = cred.secret_env
        data["credentials"].append(
            {
                "name": name,
                "profile": cred.profile,
                "secret_env": secret_env,
                "source": "env" if secret_env else "inline",
                "env_var_present": bool(os.environ.get(secret_env)) if secret_env else False,
            }
        )


@frappe.whitelist()
def reload_policy_action() -> dict[str, Any]:
    """Evict *this worker's* cached policy. Per-process only — not fleet-wide."""
    frappe.only_for(_ROLE)
    lib, lib_err = _load_library()
    if lib is None:
        return {"ok": False, "message": lib_err}

    app_policy = lib["app_policy"]
    site: str = frappe.local.site
    was_loaded = site in app_policy._policy_cache
    app_policy.reload_policy(site)
    return {
        "ok": True,
        "message": (
            f"Evicted this worker's cached policy for {site} "
            f"(was {'loaded' if was_loaded else 'not loaded'}). "
            "Other worker processes reload on their next policy miss or restart."
        ),
    }


@frappe.whitelist()
def validate_config() -> dict[str, Any]:
    """Load the policy fresh with strict checks (verifies each secret_env is set)."""
    frappe.only_for(_ROLE)
    lib, lib_err = _load_library()
    if lib is None:
        return {"ok": False, "message": lib_err, "problems": []}

    cofferdam = lib["cofferdam"]
    app_policy = lib["app_policy"]
    site: str = frappe.local.site
    display_path = app_policy.policy_path(site)
    abs_path = _abs_policy_path()

    if not os.path.exists(abs_path):
        return {"ok": False, "message": f"No policy file at {display_path}.", "problems": []}

    try:
        cofferdam.load_policy(abs_path, strict=True)
    except Exception as exc:
        problems = list(getattr(exc, "problems", None) or [])
        return {"ok": False, "message": f"Validation failed: {exc}", "problems": problems}

    return {"ok": True, "message": f"Policy at {display_path} is valid (strict).", "problems": []}


@frappe.whitelist()
def dry_run_email(recipient: str) -> dict[str, Any]:
    """Evaluate what cofferdam would do with an email to *recipient*.

    Pure decision — no Email Queue row is inserted. Mirrors the hook in
    cofferdam_app.mail, including the PRODUCTION pass-through short-circuit.
    """
    frappe.only_for(_ROLE)
    recipient = (recipient or "").strip()
    if not recipient:
        return {"ok": False, "message": "Provide a recipient address."}

    lib, lib_err = _load_library()
    if lib is None:
        return {"ok": False, "message": lib_err}

    app_policy = lib["app_policy"]
    site: str = frappe.local.site
    policy = app_policy.get_policy(site)
    if policy is None:
        return {
            "ok": False,
            "message": "No policy loaded (missing or invalid policy file). "
            "Email would be blocked (fail-closed).",
        }

    from cofferdam.mail import check_recipient, decorate_email
    from cofferdam.models import Environment

    if policy.environment is Environment.PRODUCTION:
        return {
            "ok": True,
            "recipient": recipient,
            "environment": "production",
            "allowed": True,
            "redirect_to": None,
            "reason_code": "production_passthrough",
            "decorated_subject": None,
            "message": (
                f"PRODUCTION: email to {recipient} passes through "
                "unmodified (no interception)."
            ),
        }

    decision = check_recipient(policy, recipient=recipient)
    decorated_subject, _ = decorate_email("Test subject", "Test body", policy=policy)
    if not decision.allowed:
        msg = (
            f"BLOCKED: {recipient} (env={policy.environment.value}, "
            f"reason={decision.reason_code})."
        )
    elif decision.redirect_to:
        msg = f"REDIRECTED: {recipient} → {decision.redirect_to} (env={policy.environment.value})."
    else:
        msg = f"ALLOWED: {recipient} (env={policy.environment.value})."

    return {
        "ok": True,
        "recipient": recipient,
        "environment": policy.environment.value,
        "allowed": decision.allowed,
        "redirect_to": decision.redirect_to,
        "reason_code": decision.reason_code,
        "decorated_subject": decorated_subject,
        "message": msg,
    }


@frappe.whitelist()
def dry_run_webhook(url: str) -> dict[str, Any]:
    """Evaluate what cofferdam would do with a webhook delivery to *url*.

    Pure decision — no Webhook Request Log row is inserted. Mirrors the hook in
    cofferdam_app.webhooks, including the PRODUCTION pass-through short-circuit.
    """
    frappe.only_for(_ROLE)
    url = (url or "").strip()
    if not url:
        return {"ok": False, "message": "Provide a webhook URL."}

    lib, lib_err = _load_library()
    if lib is None:
        return {"ok": False, "message": lib_err}

    app_policy = lib["app_policy"]
    site: str = frappe.local.site
    policy = app_policy.get_policy(site)
    if policy is None:
        return {
            "ok": False,
            "message": "No policy loaded (missing or invalid policy file). "
            "Webhook would be blocked (fail-closed).",
        }

    from urllib.parse import urlparse

    from cofferdam.models import Environment

    if policy.environment is Environment.PRODUCTION:
        return {
            "ok": True,
            "url": url,
            "environment": "production",
            "allowed": True,
            "reason_code": "production_passthrough",
            "message": f"PRODUCTION: webhook to {url} passes through unmodified (no interception).",
        }

    host = urlparse(url).hostname or ""
    # Constants mirror cofferdam_app.webhooks.
    decision = policy.decide(
        integration="frappe_webhooks",
        kind="webhook",
        operation="deliver",
        method="POST",
        host=host,
    )
    verdict = "ALLOWED" if decision.allowed else "BLOCKED"
    return {
        "ok": True,
        "url": url,
        "host": host,
        "environment": policy.environment.value,
        "allowed": decision.allowed,
        "reason_code": decision.reason_code,
        "message": (
            f"{verdict}: {url} (host={host or '?'}, "
            f"env={policy.environment.value}, reason={decision.reason_code})."
        ),
    }

File: cofferdam_app/cofferdam_app/page/cofferdam_status/cofferdam_status.js
// Cofferdam Status — read-only operational dashboard.
// All data is assembled server-side (cofferdam_status.py::get_data); this file
// only renders and wires the action buttons. See docs/status-page-plan.md.

frappe.pages["cofferdam-status"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Cofferdam Status"),
		single_column: true,
	});
	wrapper.cofferdam_status = new CofferdamStatus(page);
};

frappe.pages["cofferdam-status"].on_page_show = function (wrapper) {
	// Refresh on revisit so the view is never stale after a policy edit.
	if (wrapper.cofferdam_status) {
		wrapper.cofferdam_status.refresh();
	}
};

const METHOD = "cofferdam_app.cofferdam_app.page.cofferdam_status.cofferdam_status";

class CofferdamStatus {
	constructor(page) {
		this.page = page;
		this._inject_styles();
		this._build_layout();
		this._build_buttons();
		// The initial data load is driven by on_page_show, which Frappe fires
		// on first display as well as on every revisit. Refreshing here too
		// would double-fire load-time messages, so we intentionally do not.
	}

	_inject_styles() {
		if (document.getElementById("cds-styles")) return;
		const css = `
			.cds-banner { padding: 14px 18px; border-radius: 6px; color: #fff;
				margin-bottom: 14px; font-size: 15px; }
			.cds-banner .cds-env { font-size: 20px; font-weight: 700; letter-spacing: .04em; }
			.cds-banner .cds-sub { opacity: .92; margin-top: 3px; }
			.cds-green { background: #1f9d55; } .cds-amber { background: #c98a00; }
			.cds-blue { background: #2b6cb0; } .cds-red { background: #c53030; }
			.cds-section { margin-bottom: 22px; }
			.cds-section h4 { margin-bottom: 8px; }
			.cds-messages { border: 1px solid var(--border-color, #d1d8dd); border-radius: 6px;
				background: var(--fg-color, #fff); margin-bottom: 16px; }
			.cds-messages-head { display: flex; justify-content: space-between; align-items: center;
				padding: 6px 10px; border-bottom: 1px solid var(--border-color, #d1d8dd);
				font-weight: 600; font-size: 12px; text-transform: uppercase; letter-spacing: .05em; }
			.cds-messages-body { max-height: 190px; overflow-y: auto; padding: 4px 0; }
			.cds-msg { padding: 3px 12px; font-family: var(--font-stack-monospace, monospace);
				font-size: 12px; border-left: 3px solid transparent; }
			.cds-msg .cds-ts { color: var(--text-muted, #8d99a6); margin-right: 8px; }
			.cds-info { border-left-color: #2b6cb0; }
			.cds-warn { border-left-color: #c98a00; background: rgba(201,138,0,.06); }
			.cds-error { border-left-color: #c53030; background: rgba(197,48,48,.06); }
			.cds-muted { color: var(--text-muted, #8d99a6); }
			.cds-yes { color: #1f9d55; font-weight: 600; }
			.cds-no { color: #c53030; font-weight: 600; }
		`;
		const style = document.createElement("style");
		style.id = "cds-styles";
		style.textContent = css;
		document.head.appendChild(style);
	}

	_build_layout() {
		this.messages = [];
		const $main = $(this.page.main);
		$main.empty();

		this.$messages_body = $('<div class="cds-messages-body"></div>');
		const $messages = $(`
			<div class="cds-messages">
				<div class="cds-messages-head">
					<span>${__("Messages")}</span>
				</div>
			</div>`);
		const $clear = $(`<button class="btn btn-xs btn-default">${__("Clear")}</button>`);
		$clear.on("click", () => this._clear_messages());
		$messages.find(".cds-messages-head").append($clear);
		$messages.append(this.$messages_body);
		$main.append($messages);

		this.$banner = $('<div class="cds-banner cds-blue"></div>');
		$main.append(this.$banner);

		this.$body = $('<div class="cds-body"></div>');
		$main.append(this.$body);
	}

	_build_buttons() {
		this.page.set_primary_action(__("Refresh"), () => this.refresh(), "refresh");
		this.page.add_inner_button(__("Reload Policy"), () => this._reload_policy());
		this.page.add_inner_button(__("Validate Config"), () => this._validate_config());
		this.page.add_inner_button(__("Dry-run Email Check"), () => this._dry_run_email(), __("Dry-run Checks"));
		this.page.add_inner_button(__("Dry-run Webhook Check"), () => this._dry_run_webhook(), __("Dry-run Checks"));
		this.page.add_inner_button(__("Check"), () => this._metadata_check(), __("Metadata Check-Fix"));
		this.page.add_inner_button(__("Fix"), () => this._metadata_fix(), __("Metadata Check-Fix"));
	}

	// ---- Message pane -------------------------------------------------------

	log(severity, text, show_alert = true) {
		const ts = frappe.datetime.now_time();
		this.messages.unshift({ severity, text, ts });
		this.messages = this.messages.slice(0, 50);
		this._render_messages();
		if (!show_alert) return;
		// Mirror to a transient toast so the action feels responsive.
		const indicator = severity === "error" ? "red" : severity === "warn" ? "orange" : "green";
		frappe.show_alert({ message: text, indicator });
	}

	_render_messages() {
		this.$messages_body.empty();
		if (!this.messages.length) {
			this.$messages_body.append(`<div class="cds-msg cds-muted">${__("No messages.")}</div>`);
			return;
		}
		for (const m of this.messages) {
			const $m = $(`<div class="cds-msg cds-${m.severity}"></div>`);
			$m.append($('<span class="cds-ts"></span>').text(m.ts));
			$m.append(document.createTextNode(m.text));
			this.$messages_body.append($m);
		}
	}

	_clear_messages() {
		this.messages = [];
		this._render_messages();
	}

	// ---- Data load ----------------------------------------------------------

	refresh() {
		frappe.call({
			method: `${METHOD}.get_data`,
			freeze: true,
			freeze_message: __("Loading Cofferdam status…"),
			callback: (r) => {
				if (r.message) this.render(r.message);
			},
		});
	}

	render(d) {
		this._render_banner(d);
		this.$body.empty();
		this.$body.append(this._section_status(d));
		this.$body.append(this._section_mail(d));
		this.$body.append(this._section_effects(d));
		this.$body.append(this._section_integrations(d));
		this.$body.append(this._section_credentials(d));
		this.$body.append(this._section_coverage(d));

		// Surface load-time findings in the message pane.
		if (!d.library_installed) {
			this.log("error", __("cofferdam library is NOT installed — interception is inert."));
		}
		(d.errors || []).forEach((e) => this.log("error", e));
	}

	_render_banner(d) {
		let cls = "cds-blue";
		let env = (d.environment || "").toUpperCase();
		let headline;
		if (!d.library_installed) {
			cls = "cds-red";
			headline = __("cofferdam NOT INSTALLED");
		} else if (!d.policy_file_exists || !d.environment) {
			cls = "cds-red";
			headline = __("POLICY MISSING / ERROR");
		} else if (env === "PRODUCTION") {
			cls = "cds-green";
			headline = env;
		} else if (env === "STAGING") {
			cls = "cds-amber";
			headline = env;
		} else {
			cls = "cds-blue";
			headline = env; // DEV / TEST
		}
		this.$banner.attr("class", `cds-banner ${cls}`);
		this.$banner.empty();
		this.$banner.append($('<div class="cds-env"></div>').text(headline));
		const sub = `${__("Site")}: ${d.site}   •   ${__("Policy")}: ${d.policy_file_path || "—"}`;
		this.$banner.append($('<div class="cds-sub"></div>').text(sub));
	}

	// ---- Sections -----------------------------------------------------------

	_section(title, $inner) {
		const $s = $('<div class="cds-section"></div>');
		$s.append($("<h4></h4>").text(title));
		$s.append($inner);
		return $s;
	}

	_kv_table(rows) {
		const $t = $('<table class="table table-bordered"></table>');
		const $b = $("<tbody></tbody>");
		for (const [k, v] of rows) {
			const $tr = $("<tr></tr>");
			$tr.append($("<td></td>").css("width", "260px").text(k));
			$tr.append($("<td></td>").html(v));
			$b.append($tr);
		}
		$t.append($b);
		return $t;
	}

	_section_status(d) {
		const rows = [
			[__("Library installed"), yesno(d.library_installed)],
			[__("Library version"), esc(d.library_version || __("not installed"))],
			[__("Frappe App version"), esc(d.app_version)],
			[__("Policy file path"), esc(d.policy_file_path || "—")],
			[__("Policy file present"), yesno(d.policy_file_exists)],
			[__("Last modified"), esc(d.policy_file_mtime || "—") + ` <span class="cds-muted">(${__("informational only")})</span>`],
			[
				__("Loaded in this worker"),
				yesno(d.policy_loaded_in_this_worker) +
					` <span class="cds-muted">(${__("per-process, not fleet-wide")})</span>`,
			],
			[__("Default decision"), esc(d.default_decision || "—")],
		];
		return this._section(__("Policy file & versions"), this._kv_table(rows));
	}

	_section_mail(d) {
		if (!d.mail) {
			return this._section(
				__("Mail policy"),
				$(`<div class="cds-muted">${__("No [mail] section in policy.")}</div>`)
			);
		}
		const m = d.mail;
		const rows = [
			[__("Mode"), esc(m.mode)],
			[__("Sink address"), esc(m.sink || "—")],
			[__("Allowed domains"), esc((m.allow_domains || []).join(", ") || "—")],
			[__("Decoration enabled"), yesno(m.decorate)],
		];
		return this._section(__("Mail policy"), this._kv_table(rows));
	}

	_section_effects(d) {
		const cols = [__("Kind"), __("Scope"), __("Enabled"), __("Allowed domains")];
		const rows = (d.effects || []).map((e) => [
			esc(e.kind),
			esc(e.scope),
			yesno(e.enabled),
			esc((e.allow_domains || []).join(", ") || "—"),
		]);
		return this._section(__("Effects"), this._grid(cols, rows, __("No effects rules defined.")));
	}

	_section_integrations(d) {
		const cols = [
			__("Name"), __("Kind"), __("Enabled"), __("Credential"), __("Allowed hosts"),
			__("Allowed methods"), __("Allowed operations"), __("Authorize"), __("Capture"),
		];
		const rows = (d.integrations || []).map((i) => [
			esc(i.name),
			esc(i.kind),
			yesno(i.enabled),
			esc(i.credential || "—"),
			esc((i.allowed_hosts || []).join(", ") || "—"),
			esc((i.allowed_methods || []).join(", ") || "—"),
			esc((i.allowed_operations || []).join(", ") || "—"),
			yesno(i.allow_authorize),
			yesno(i.allow_capture),
		]);
		return this._section(__("Integrations"), this._grid(cols, rows, __("No integrations defined.")));
	}

	_section_credentials(d) {
		const cols = [__("Name"), __("Profile"), __("Source"), __("Env var"), __("Env var present?")];
		const rows = (d.credentials || []).map((c) => [
			esc(c.name),
			esc(c.profile),
			esc(c.source),
			esc(c.secret_env || "—"),
			c.source === "inline"
				? `<span class="cds-muted">${__("inline secret")}</span>`
				: yesno(c.env_var_present),
		]);
		return this._section(
			__("Credentials (names only — no secret values)"),
			this._grid(cols, rows, __("No credentials defined."))
		);
	}

	_section_coverage(d) {
		const cols = [__("Outbound path"), __("Intercepted?"), __("Note")];
		const rows = (d.coverage || []).map((c) => [
			esc(c.path),
			c.intercepted ? `<span class="cds-yes">✅</span>` : `<span class="cds-no">❌</span>`,
			esc(c.note),
		]);
		return this._section(__("Interception coverage"), this._grid(cols, rows, ""));
	}

	_grid(cols, rows, empty_text) {
		const $t = $('<table class="table table-bordered"></table>');
		const $h = $("<thead></thead>");
		const $hr = $("<tr></tr>");
		cols.forEach((c) => $hr.append($("<th></th>").text(c)));
		$h.append($hr);
		$t.append($h);
		const $b = $("<tbody></tbody>");
		if (!rows.length) {
			const $tr = $("<tr></tr>");
			$tr.append(
				$("<td></td>")
					.attr("colspan", cols.length)
					.addClass("cds-muted")
					.text(empty_text || __("None."))
			);
			$b.append($tr);
		} else {
			for (const cells of rows) {
				const $tr = $("<tr></tr>");
				cells.forEach((cell) => $tr.append($("<td></td>").html(cell)));
				$b.append($tr);
			}
		}
		$t.append($b);
		return $t;
	}

	// ---- Actions ------------------------------------------------------------

	_reload_policy() {
		frappe.call({
			method: `${METHOD}.reload_policy_action`,
			callback: (r) => {
				const m = r.message || {};
				this.log(m.ok ? "warn" : "error", m.message || __("Reload failed."));
			},
		});
	}

	_validate_config() {
		frappe.call({
			method: `${METHOD}.validate_config`,
			callback: (r) => {
				const m = r.message || {};
				this.log(m.ok ? "info" : "error", m.message || __("Validation failed."));
				(m.problems || []).forEach((p) => this.log("error", `  • ${p}`));
			},
		});
	}

	_dry_run_email() {
		frappe.prompt(
			[
				{ fieldname: "recipient", fieldtype: "Data", label: __("Recipient address"), reqd: 1 },
				{
					fieldtype: "HTML",
					options: `<p class="text-muted">${__(
						"No email will be sent. This only checks how Cofferdam rules would handle the recipient."
					)}</p>`,
				},
			],
			(v) => {
				frappe.call({
					method: `${METHOD}.dry_run_email`,
					args: { recipient: v.recipient },
					callback: (r) => {
						const m = r.message || {};
						const result = [m.message || __("No result.")];
						if (m.ok && m.decorated_subject) {
							result.push(`${__("Decorated subject")}: ${m.decorated_subject}`);
						}
						this.log(this._decision_severity(m), result.join(" "), false);
						this._show_result_output(__("Dry-run Email Check"), result.join("\n"), m.ok);
					},
				});
			},
			__("Dry-run Email Check"),
			__("Evaluate")
		);
	}

	_dry_run_webhook() {
		frappe.prompt(
			[{ fieldname: "url", fieldtype: "Data", label: __("Webhook URL"), reqd: 1 }],
			(v) => {
				frappe.call({
					method: `${METHOD}.dry_run_webhook`,
					args: { url: v.url },
					callback: (r) => {
						const m = r.message || {};
						const result = m.message || __("No result.");
						this.log(this._decision_severity(m), result, false);
						this._show_result_output(__("Dry-run Webhook Check"), result, m.ok);
					},
				});
			},
			__("Dry-run Webhook Check"),
			__("Evaluate")
		);
	}

	_metadata_check() {
		this._run_metadata_action("metadata_check", __("Metadata Check"), "info");
	}

	_metadata_fix() {
		frappe.confirm(
			__("Fix will permanently delete only dangling Accounting Dimensions, Custom Fields, and Property Setters found by the audit. Continue?"),
			() => this._run_metadata_action("metadata_fix", __("Metadata Fix"), "warn"),
			() => {}
		);
	}

	_run_metadata_action(action, title, success_severity) {
		frappe.call({
			method: `${METHOD}.${action}`,
			freeze: true,
			freeze_message: action === "metadata_fix" ? __("Repairing metadata…") : __("Checking metadata…"),
			callback: (r) => {
				const m = r.message || {};
				this.log(m.ok ? success_severity : "error", m.message || __("No result."));
				this._show_result_output(title, m.stdout || m.message || __("No output."), m.ok);
				if (m.ok && action === "metadata_fix") this.refresh();
			},
		});
	}

	_show_result_output(title, output, ok) {
		frappe.msgprint({
			title,
			indicator: ok ? "green" : "red",
			message: `<pre style="max-height: 420px; overflow: auto; white-space: pre-wrap; margin: 0;">${esc(output)}</pre>`,
		});
	}

	_decision_severity(m) {
		if (!m.ok) return "error";
		if (m.allowed === false) return "warn";
		return "info";
	}
}

function esc(v) {
	return frappe.utils.escape_html(v == null ? "" : String(v));
}

function yesno(v) {
	return v
		? `<span class="cds-yes">${__("Yes")}</span>`
		: `<span class="cds-no">${__("No")}</span>`;
}
